File: FunctionalProgramming/FunctionalProgramming.CSharp/Recursion/RecursiveTrees.cs
// from https://www.hackerrank.com/challenges/fractal-trees

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using FunctionalProgramming.Common;
using NUnit.Framework;

namespace FunctionalProgramming.CSharp.Recursion
{
    public class RecursiveTreesTask : ITask
    {
        public RecursiveTreesTask(TextReader inputReader, TextWriter outputWriter)
        {
            _inputReader = inputReader;
            _outputWriter = outputWriter;
        }

        public int Execute(String[] args)
        {
            Show(BuildTree(Int32.Parse(_inputReader.ReadLine())));
            return 0;
        }

        private LinkedList<IList<Int32>> BuildTree(Int32 count)
        {
            return BuildTree(1, count, SegmentStartSize, new LinkedList<IList<Int32>>());
        }

        private LinkedList<IList<Int32>> BuildTree(Int32 level, Int32 count, Int32 size, LinkedList<IList<Int32>> storage)
        {
            if (level > count)
                return storage;
            return BuildTree(level + 1, count, size / 2, BuildSlantingSegments(size, BuildVerticalSegment(size, storage)));
        }

        private LinkedList<IList<Int32>> BuildVerticalSegment(Int32 size, LinkedList<IList<Int32>> storage)
        {
            return Enumerable.Range(1, size).Aggregate(storage, (acc, _) => storage.AddFirst(storage.Count > 0 ? storage.First.Value : new[] {ColumnStart}).List);
        }

        private LinkedList<IList<Int32>> BuildSlantingSegments(Int32 size, LinkedList<IList<Int32>> storage)
        {
            return BuildSlantingSegments(size, storage, storage.First.Value);
        }

        private LinkedList<IList<Int32>> BuildSlantingSegments(Int32 size, LinkedList<IList<Int32>> storage, IList<Int32> sourceValue)
        {
            return Enumerable.Range(1, size).Aggregate(storage, (acc, delta) => storage.AddFirst(GenerateSlantingSegmentsRow(sourceValue, delta)).List);
        }

        private IList<Int32> GenerateSlantingSegmentsRow(IList<Int32> sourceValue, Int32 delta)
        {
            return sourceValue.Aggregate(new List<Int32>(), (acc, value) =>
            {
                acc.AddRange(new[] {value - delta, value + delta});
                return acc;
            });
        }

        private LinkedList<IList<Int32>> ShowEmptyLines(LinkedList<IList<Int32>> storage)
        {
            // TODO (std_string) : need something such List.iter from F#
            Enumerable.Range(1, RowCount - storage.Count).Aggregate(new Object(), (acc, _) =>
            {
                _outputWriter.WriteLine(new String(EmptyChar, ColumnCount));
                return acc;
            });
            return storage;
        }

        private void ShowLine(IList<Int32> colunms)
        {
            Enumerable.Range(1, ColumnCount).Aggregate(0, (acc, index) =>
            {
                if (acc == colunms.Count)
                {
                    _outputWriter.Write(EmptyChar);
                    return acc;
                }
                _outputWriter.Write(colunms[acc] == index ? LineChar : EmptyChar);
                return colunms[acc] == index ? acc + 1 : acc;
            });
            _outputWriter.WriteLine();
        }

        private void ShowLines(LinkedList<IList<Int32>> storage)
        {
            // TODO (std_string) : need something such List.iter from F#
            storage.Aggregate(new Object(), (acc, columns) =>
            {
                ShowLine(columns);
                return acc;
            });
        }

        private void Show(LinkedList<IList<Int32>> storage)
        {
            ShowLines(ShowEmptyLines(storage));
        }

        private readonly TextReader _inputReader;

        private readonly TextWriter _outputWriter;

        private const Int32 RowCount = 63;

        private const Int32 ColumnCount = 100;

        private const Char EmptyChar = '_';

        private const Char LineChar = '1';

        private const Int32 ColumnStart = 50;

        private const Int32 SegmentStartSize = 16;
    }

    [TestFixture]
    public class RecursiveTreesTests
    {
        [TestCase("1", "Output00.txt")]
        [TestCase("2", "Output01.txt")]
        [TestCase("3", "Output02.txt")]
        [TestCase("4", "Output03.txt")]
        [TestCase("5", "Output04.txt")]
        public void Execute(String input, String expectedOutputFile)
        {
            String expectedOutput = File.ReadAllText(Path.Combine(RootDirectory, expectedOutputFile));
            TaskExecuter.Execute((inputReader, outputWriter) => new RecursiveTreesTask(inputReader, outputWriter), input, expectedOutput);
        }

        private const String RootDirectory = ".//TestCases//Recursion//RecursiveTrees";
    }
}

File: ProblemSolving/FunctionalProgramming.FSharp/Recursion/RecursiveTrees.fs
// from https://www.hackerrank.com/challenges/fractal-trees

namespace FunctionalProgramming.FSharp.Recursion

open NUnit.Framework
open ProblemSolving.Common
open System.IO

type RecursiveTreesTask(input: TextReader, output: TextWriter) =

    [<Literal>]
    let RowCount = 63

    [<Literal>]
    let ColumnCount = 100

    [<Literal>]
    let EmptyChar = '_'

    [<Literal>]
    let LineChar = '1'

    [<Literal>]
    let ColumnStart = 50

    [<Literal>]
    let SegmentStartSize = 16

    let buildVerticalSegment (size: int) (storage: int list list) =
        match storage with
        | [] -> [ColumnStart - 1] |> List.replicate size
        | head :: _ -> storage |> List.append (head |> List.replicate size)

    let buildSlantingSegments (size: int) (storage: int list list) =
        storage |> List.append [for delta in size .. -1 .. 1 -> storage |>  List.head |> List.collect (fun column -> [column - delta; column + delta])]

    let rec buildTree (level: int) (count: int) (size: int) (storage: int list list) =
        match level with
        | _ when level > count -> storage
        | _ -> storage |> buildVerticalSegment size |> buildSlantingSegments size |> buildTree (level + 1) count (size / 2)

    let buildTreeStart (count: int) =
        buildTree 1 count SegmentStartSize []

    let showEmptyLines (storage: int list list) =
        seq {1 .. (RowCount - List.length storage)} |> Seq.iter (fun _ -> new System.String(EmptyChar, ColumnCount) |> output.WriteLine)
        storage

    let rec showLine (index: int) (columns: int list) =
        match index with
        | ColumnCount -> "" |> output.WriteLine
        | _ ->
            match columns with
            | [] -> EmptyChar |> output.Write; showLine (index + 1) columns
            | column :: rest when index = column -> LineChar |> output.Write; showLine (index + 1) rest
            | _ -> EmptyChar |> output.Write; showLine (index + 1) columns

    let showStorage (storage: int list list) =
        storage |> List.iter (fun columns -> showLine 0 columns)

    interface ITask with
        member this.Execute(argv: string[]) =
            input.ReadLine() |> int |> buildTreeStart |> showEmptyLines |> showStorage
            0

[<TestFixture>]
type RecursiveTreesTests() =

    [<Literal>]
    let RootDirectory = ".//TestCases//Recursion//RecursiveTrees"

    [<TestCase("1", "Output00.txt")>]
    [<TestCase("2", "Output01.txt")>]
    [<TestCase("3", "Output02.txt")>]
    [<TestCase("4", "Output03.txt")>]
    [<TestCase("5", "Output04.txt")>]
    member public this.Execute(input: string, expectedOutputFile: string) =
        let expectedOutput = File.ReadAllText(Path.Combine(RootDirectory, expectedOutputFile))
        TaskExecutor.Execute((fun reader writer -> new RecursiveTreesTask(reader, writer) :> ITask), input, expectedOutput)